File: scripts/py/bible_audio_repair_queue.py
"""The chapters whose recording needs doing again, found rather than remembered.

A chapter belongs here when both things are true of it: its sound was written
before the reading could sound out a name it does not know, and the reading as
it was then would have dropped a word of that chapter.  Either half alone names
the wrong set - the first alone takes in hundreds of chapters that happen to hold
no hard name, and the second alone cannot tell a chapter already done again from
one nobody has touched.

The set narrows itself as the work is done.  A chapter recorded again gets sound
newer than the door, so it fails the first test the next time this is asked and
is never offered twice.  That is also what makes the asking cheaper each night:
only the chapters still holding old sound are read at all.

Takes the path of a JSON file holding {"root": <folder of chapter folders>,
"before_second": <the second the door was made>}, and prints one JSON line
naming the queue.  It reads only; nothing is recorded or removed.
"""

import json
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from bible_audio_words_dropped import (  # noqa: E402
    chapter_texts,
    dropped_in,
    g2p_as_it_was,
)


def sound_before(folder, before_second):
    """Whether every numbered piece of this chapter was written before that second.

    Every piece is asked about rather than the newest one, because a chapter can
    hold a joined whole-chapter file written long after its pieces, and that file
    says nothing about which reading spoke them.  A folder holding no numbered
    piece at all is not a recorded chapter and is answered no.
    """
    pieces = 0
    for name in os.listdir(folder):
        if not name.endswith(".mp3"):
            continue
        stem = name[: -len(".mp3")]
        if not stem.isdigit():
            continue
        pieces += 1
        if os.path.getmtime(os.path.join(folder, name)) >= before_second:
            return False
    return pieces > 0


def main(args_path):
    with open(args_path, encoding="utf-8") as fh:
        args = json.load(fh)
    root = args["root"]
    before_second = args["before_second"]

    names = sorted(
        n for n in os.listdir(root) if os.path.isdir(os.path.join(root, n))
    )
    old_sound = [n for n in names if sound_before(os.path.join(root, n), before_second)]

    g2p = g2p_as_it_was()
    queue = []
    words_dropped = 0
    for name in old_sound:
        texts = chapter_texts(os.path.join(root, name))
        if not texts:
            continue
        lost, _words = dropped_in(g2p, texts)
        if lost:
            queue.append(name)
            words_dropped += sum(lost.values())

    print(
        json.dumps(
            {
                "root": root,
                "before_second": before_second,
                "chapters_all": len(names),
                "chapters_sound_before_door": len(old_sound),
                "queue": queue,
                "words_dropped": words_dropped,
            },
            ensure_ascii=False,
        )
    )


if __name__ == "__main__":
    main(sys.argv[1])

File: js/bible_audio_repair_queue.mjs
import { arguments_assert } from "./arguments_assert.mjs";
import { bible_audio_reading_changed_queue } from "./bible_audio_reading_changed_queue.mjs";
import { not } from "./not.mjs";
import { property_get } from "./property_get.mjs";
import { list_concat } from "./list_concat.mjs";
export async function bible_audio_repair_queue(bible_folder) {
  "$plain bible_folder";
  "The chapters of a recorded bible whose sound needs making again, because the reading of the day it was spoken says something the reading now does not - a word left silent, or a word said differently.";
  "★ THE TEST IS THAT THE RECORDING SAYS SOMETHING WRONG, NOT THAT IT WENT SILENT. Silence was only ever the half of that which was easy to see: a name the old reading had never heard of came back with no sounds at all, while a name it held a wrong answer for came back sounding wrong, which is not silence and went unnoticed. Measured on the fifth of September 2026 that was 507 chapters caught against 1086 at fault, so more than half the broken recordings were being called good.";
  "★ THE SMALL WORDS ARE BROKEN TOO, WHICH IS WHY THE WIDER HALF IS NOT A MATTER OF TASTE. A reading that cannot sound out a name has no first sound for the word before it to read, so the rule that says the before a vowel and the before a consonant fired on nothing and fired wrongly. Counted over every chapter, that is 208 thes and 616 tos said wrongly - each one of them beside a name, and every single one of them right in the reading as it is now.";
  "★ IT FINDS ITS OWN SET, WHICH IS THE WHOLE REASON IT IS A COMMAND AND NOT A LIST. A pass over a thousand chapters does not fit in one night, so the same question gets asked again the next night and the night after. A list handed in would name chapters already done and spend a night doing them twice; asked afresh, a chapter that has been recorded again holds newer sound and simply is not in the answer any more.";
  "★ THE SILENT HALF IS OFFERED FIRST BECAUSE IT IS THE WORSE FAULT. A chapter missing a word outright says less than it should; a chapter saying a name wrongly says the wrong thing but says it all. A night that runs out part way through then spends itself on the worse half, which is the only ordering that makes a partial pass worth as much as it can be.";
  "★ ASKING COSTS ABOUT EIGHT MINUTES, BECAUSE EVERY CHAPTER IS NOW READ TWICE. That is paid once a night against a window of nine hours, and it buys the only question that can tell a mended chapter from a broken one - the older asking read each chapter once and could not see a name said wrongly at all.";
  arguments_assert(arguments, 1);
  let reported = await bible_audio_reading_changed_queue(bible_folder, "");
  if (not(reported)) {
    return null;
  }
  let dropped = property_get(reported, "queue_dropped");
  let changed = property_get(reported, "queue_changed_only");
  let queue = list_concat(dropped, changed);
  let wider = {
    ...reported,
    queue,
  };
  return wider;
}
